### PR TITLE
Extend_ssp_authsources_template

### PR DESCRIPTION
**Pull request by ioigoume**
Extend SSP authsources.php.j2 template in order to support raw php

### CODE REVIEW

File: roles/ssp/templates/config/authsources-1.14.php.j2 [hunk 1]
```diff
@@ -4,6 +4,10 @@
  * {{ ansible_managed }}
  */
 
+{% if ssp_authsources_function_raw is defined %}
+{{ ssp_authsources_function_raw }}
```
Comment (NicolasLiampotis) [inline roles/ssp/templates/config/authsources-1.14.php.j2:8]: rename to ssp_authsources_preamble
and add it to defaults/main.yml (in comments)

```
# Added before the config[] in SimpleSAMLphp authentication sources
#ssp_authsources_preamble: |
#    function foo() {
```

File: roles/ssp/templates/config/authsources-1.14.php.j2 [hunk 2]
```diff
@@ -25,7 +29,7 @@ $config = array(
 {% if authsource.idp is defined %}
         // The entity ID of the IdP this should SP should contact.
         // Can be NULL/unset, in which case the user will be shown a list of available IdPs.
-        'idp' => '{{ authsource.idp }}',
+        'idp' => {{ authsource.idp }},
```
Comment (NicolasLiampotis) [inline roles/ssp/templates/config/authsources-1.14.php.j2:32]: Introduce `authsource.idp_function`

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]

File: roles/ssp/templates/config/authsources-1.14.php.j2 (hunk 1 above)
Comment (ioigoume) [reply roles/ssp/templates/config/authsources-1.14.php.j2]: Done

File: roles/ssp/templates/config/authsources-1.14.php.j2 (hunk 2 above)
Comment (ioigoume) [reply roles/ssp/templates/config/authsources-1.14.php.j2]: Done

File: roles/ssp/templates/config/authsources-1.14.php.j2 [hunk 3]
```diff
@@ -22,7 +26,10 @@ $config = array(
         // Can be NULL/unset, in which case an entity ID is generated based on the metadata URL.
         'entityID' => '{{ authsource.entity_id }}',
 {% endif %}
-{% if authsource.idp is defined %}
+{% if authsource.idp_function is defined %}
+        // The entity ID of the IdP constructed dynamically using a function.
```
Comment (NicolasLiampotis) [inline roles/ssp/templates/config/authsources-1.14.php.j2:30]: Replace with:
// The function returning the entity ID of the IdP this SP should contact.
// Can be NULL/unset, in which case the user will be shown a list of available IdPs.

File: roles/ssp/templates/config/authsources-1.17.php.j2 [hunk 4]
```diff
@@ -22,7 +26,10 @@ $config = [
         // Can be NULL/unset, in which case an entity ID is generated based on the metadata URL.
         'entityID' => '{{ authsource.entity_id }}',
 {% endif %}
-{% if authsource.idp is defined %}
+{% if authsource.idp_function is defined %}
+        // The entity ID of the IdP constructed dynamically using a function.
```
Comment (NicolasLiampotis) [inline roles/ssp/templates/config/authsources-1.17.php.j2:30]: Replace with
// The function returning the entity ID of the IdP this SP should contact.
// Can be NULL/unset, in which case the user will be shown a list of available IdPs.

File: roles/ssp/templates/config/authsources-1.18.php.j2 [hunk 5]
```diff
@@ -22,7 +26,10 @@ $config = [
         // Can be NULL/unset, in which case an entity ID is generated based on the metadata URL.
         'entityID' => '{{ authsource.entity_id }}',
 {% endif %}
-{% if authsource.idp is defined %}
+{% if authsource.idp_function is defined %}
+        // The entity ID of the IdP constructed dynamically using a function.
```
Comment (NicolasLiampotis) [inline roles/ssp/templates/config/authsources-1.18.php.j2:30]: Replace with:
// The function returning the entity ID of the IdP this SP should contact.
// Can be NULL/unset, in which case the user will be shown a list of available IdPs.

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]
We need to document the `idp` and the `idp_function` options (in comments) in defaults/main.yml:

- name: default-sp
    # Uncomment to specify the entity ID of the IdP this should SP should
    # contact. If this is not set, the user will be shown a list of available
    # IdPs.
    # idp: "https://idp.example.org
    # Uncomment to specify the function returning the entity ID of the IdP
    # this SP should contact. When NULL the user will be shown a list of
    # available IdPs.
    # idp_function: "findIdP()"

File: roles/ssp/templates/config/authsources-1.14.php.j2 (hunk 3 above)
Comment (ioigoume) [reply roles/ssp/templates/config/authsources-1.14.php.j2:30]: Done

File: roles/ssp/templates/config/authsources-1.18.php.j2 (hunk 5 above)
Comment (ioigoume) [reply roles/ssp/templates/config/authsources-1.18.php.j2:30]: Done

File: roles/ssp/templates/config/authsources-1.17.php.j2 (hunk 4 above)
Comment (ioigoume) [reply roles/ssp/templates/config/authsources-1.17.php.j2:30]: Done

File: roles/ssp/defaults/main.yml [hunk 6]
```diff
@@ -371,6 +382,8 @@ ssp_authsources_saml:
     #      class: "saml:PersistentNameID2TargetedID"
     #      attribute: "eduPersonTargetedID"
     #      nameId: no
+    #idp_function: foo()
+    # Optional function which will construct the IdP entity ID
```
Comment (NicolasLiampotis) [inline roles/ssp/defaults/main.yml:386]: Remove duplicate definition of `idp_function` param

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]
Remove duplicate definition of `idp_function` param from `defaults/main.yml`

File: roles/ssp/defaults/main.yml (hunk 6 above)
Comment (ioigoume) [reply roles/ssp/defaults/main.yml]: Done

**Review by NicolasLiampotis — APPROVED** [review]
LGTM

**Pull request merged**